presto-main/src/main/java/com/facebook/presto/ExceededMemoryLimitException.java

```diff
@@ -35,8 +38,13 @@ public static ExceededMemoryLimitException exceededGlobalTotalLimit(DataSize max
         return new ExceededMemoryLimitException(EXCEEDED_GLOBAL_MEMORY_LIMIT, format("Query exceeded distributed total memory limit of %s defined at the %s", maxMemory, limitSource));
     }
 
-    public static ExceededMemoryLimitException exceededLocalUserMemoryLimit(DataSize maxMemory, String additionalFailureInfo)
+    public static ExceededMemoryLimitException exceededLocalUserMemoryLimit(
+            DataSize maxMemory,
+            String additionalFailureInfo,
+            boolean heapDumpOnExceededMemoryLimitEnabled,
+            Optional<String> heapDumpFilePath)
     {
+        performHeapDumpIfEnabled(heapDumpOnExceededMemoryLimitEnabled, heapDumpFilePath);
         return new ExceededMemoryLimitException(EXCEEDED_LOCAL_MEMORY_LIMIT,
```

Comment:
It's kinda weird having a dump inside an exception and do it synchronically.... Can we at the callsite, when memory exceeds, we throw the error and asynchronically use an executor to do the dump?

Reply:
It's important the heap dump be triggered synchronously with the out of memory error because the purpose of the entire thing is capture the state of the heap when the error occurs. If it's asynchronous it allows time for queries to fail and cleanup to occur both in this query and in other queries that are running concurrently.

WDYT?

Reply:
If that is the case, shall we add comments to this method (and other's in this class) to indicate this needed and only for debugging purpose.